### PR TITLE
add build script to compile smart contract locally using smartpy cli

### LINKED ISSUE #45: Add build script for compiling smartPy contract into michelson
Need to automate compiling the smartPy contract to support CI.

### PR DESCRIPTION
**Pull request by jakinyele**
closes #45

### CODE REVIEW

**Review by DariusParvin — APPROVED** [review]
Tested and worked successfully, just one minor suggestion.

File: build-contract.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  if [[ $OS = "ubuntu" ]]; then
+    console "Compiling on Ubuntu ($OS_VERSION)"
+    sh <(curl -s https://smartpy.io/cli/install.sh)
+  elif [[ $OS = "debian" ]]; then
+    console "Compiling on Debian ($OS_VERSION)"
+  else
+    fail "Need install steps for your OS: ($OS_VERSION)"
+  fi
+fi
```
Comment (yaymukund) [inline build-contract.sh:102]: Would this be clearer?

```bash
if ! [[ "$OS_RELEASE_ID" =~ ^(darwin|debian|ubuntu)$ ]] ; then
  fail "Need install steps for $OS_RELEASE_ID"
fi

if [[ "$OS_RELEASE_ID" = 'darwin' ]] ; then
  # 
fi

if [[ "$OS_RELEASE_ID" = 'ubuntu' ]] ; then
  # ...
else
  # ...
fi
```

File: build-contract.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+set -x
+# install the smartPy CLI v0.7.4 (if not currently present)
+sh <(curl -s https://smartpy.io/releases/20210904-98c3fb1314a5298a5000fe3801d0b57238469670/cli/install.sh)
```
Comment (yaymukund) [inline build-contract.sh:106]: When is this supposed to run? Could this be grouped with the smartypy install logic up above?

File: build-contract.sh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+# create the output dir
+mkdir -p $TEMP_DIR/
+# first let's test the script to ensure no build errors as a sanity check
+$HOME/smartpy-cli/SmartPy.sh test $SMARTPY_CONTRACT $TEMP_DIR/
```
Comment (yaymukund) [inline build-contract.sh:110]: Should this potentially early-terminate the script? (`set -e`?)

File: build-contract.sh [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+# then proceed to compile the smartPy script in the target output directory
+$HOME/smartpy-cli/SmartPy.sh compile $SMARTPY_CONTRACT $TEMP_DIR/
```
Comment (yaymukund) [inline build-contract.sh:112]: Is this always expected to live in `$HOME/smartpy-cli`?

File: build-contract.sh [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+function brew_install() {
+    console "Installing $1"
+    if brew list $1 &>/dev/null; then
+        echo "${1} is already installed"
+    else
+        brew install $1 && echo "$1 is installed"
+    fi
+}
+
+TEMP_DIR=$(shell mktemp -d tmp-XXXX)
```
Comment (yaymukund) [inline build-contract.sh:76]: I'm not on any of the supported platforms - so maybe this doesn't apply - but I get `shell: command not found`. 

Also, it looks like `mktemp` has a slightly different interface on OSX v Linux: https://unix.stackexchange.com/a/84980

File: build-contract.sh [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+function brew_install() {
+    console "Installing $1"
+    if brew list $1 &>/dev/null; then
+        echo "${1} is already installed"
+    else
+        brew install $1 && echo "$1 is installed"
+    fi
+}
```
Comment (yaymukund) [inline build-contract.sh:74]: :ok_hand:

**Review by yaymukund — CHANGES_REQUESTED** [review]
I think ideally we should do feature-detection (e.g. check for the presence of the `brew` or `apt` rather than look at `os-release`), but I don't think it's high priority.

File: build-contract.sh [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  elif [[ -f "$CENTOS_RELEASE" ]]; then
+    FAMILY="centos"
+    eval $__out="centos"
+  elif [[ -f "$REDHAT_RELEASE" ]]; then
+    FAMILY="redhat"
+    eval $__out="redhat"
+  else
+    eval $__out=`uname -s | tr '[:upper:]' '[:lower:]'`
+  fi
+}
```
[the remark this answers is not in the document]
Comment (jakinyele) [reply build-contract.sh:52]: Agreed

File: build-contract.sh [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+# Usage: ./build-contract.sh [ /optional/path/to/dir ]
+#
+
+CENTOS_RELEASE=/etc/centos-release
+REDHAT_RELEASE=/etc/redhat-release
+FEDORA_RELEASE=/etc/fedora-release
+LSB_RELEASE=/etc/lsb-release
+ORACLE_RELEASE=/etc/oracle-release
+SYSTEM_RELEASE=/etc/system-release
+DEBIAN_VERSION=/etc/debian_version
```
[the remark this answers is not in the document]
Comment (jakinyele) [reply build-contract.sh:16]: Great point. It makes sense to switch to `/etc/os-release` and thanks for the link!

File: build-contract.sh (hunk 1 above)
Comment (jakinyele) [reply build-contract.sh:102]: Yup!

File: build-contract.sh (hunk 2 above)
Comment (jakinyele) [reply build-contract.sh:106]: Supposed to run once to confirm correct smartpy install but this line can be grouped to check for an existing install first.

File: build-contract.sh (hunk 3 above)
Comment (jakinyele) [reply build-contract.sh:110]: good point. yeah, it should early terminate. will update

File: build-contract.sh (hunk 4 above)
Comment (jakinyele) [reply build-contract.sh:112]: yes, but will look at the innards of the installer to see if that can be customized.

File: build-contract.sh (hunk 5 above)
Comment (jakinyele) [reply build-contract.sh:76]: oh i see. will fix. thanks for the pointer!

File: build-contract.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (yaymukund) [reply build-contract.sh:51]: I wrote it confusingly, but I'm suggesting we do this (untested code ahead):

```rust
OS_RELEASE=/etc/os-release

get_os_release_id() {
  local __out=$1

  if [[ -f "$OS_RELEASE" ]] ; then
    eval $__out=$(awk -F= '$1=="ID" { print $2 ;}' $OS_RELEASE)
  elif [[ $OS_TYPE == "darwin"* ]] ; then
    eval $__out="darwin"
  else
    eval $__out='unknown'
  fi
}

get_os_release_version() {
  local __out=$1

  if [[ -f $OS_RELEASE ]] ; then
    eval $__out=$(awk -F= '$1=="VERSION" { print $2 ;}' $OS_RELEASE)
  elif [[ $OS_TYPE == "darwin"* ]] ; then
    eval $__out=$(uname -s | tr '[:upper:]' '[:lower:]')
  else
    eval $__out='Unknown'
  fi
}
```

**Review by yaymukund — COMMENTED** [review]
I tried running it:

- `SmartPy`'s install script expects npm to be installed. I wonder if this should be mentioned somewhere, or if we should check for it explicitly.
- I'm seeing an error compiling for some reason. It's deep within some smartpyc stack, but here it is:
    ```
    (node:235848) RangeError: Invalid count value
    at String.repeat (<anonymous>)
    at caml_str_repeat (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:103:28)
    at caml_convert_string_to_bytes (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:348:15)
    at caml_jsbytes_of_string (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:353:25)
    at caml_ml_output_bytes (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:5260:17)
    at caml_ml_output (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:5270:14)
    at output_string (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:9322:14)
    at print_endline (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:9421:7)
    at caml_call1 (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:247603:28)
    at Main (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:248536:9)
    ```
    This happens during the `SmartPy.sh test` step on L107 while trying to construct the contract_*.tar.gz.

**Review by yaymukund — APPROVED** [review]

**Comment by jakinyele** [comment]
> I tried running it:
> 
> * `SmartPy`'s install script expects npm to be installed. I wonder if this should be mentioned somewhere, or if we should check for it explicitly.
> * I'm seeing an error compiling for some reason. It's deep within some smartpyc stack, but here it is:
>   ```
>   (node:235848) RangeError: Invalid count value
>   at String.repeat (<anonymous>)
>   at caml_str_repeat (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:103:28)
>   at caml_convert_string_to_bytes (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:348:15)
>   at caml_jsbytes_of_string (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:353:25)
>   at caml_ml_output_bytes (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:5260:17)
>   at caml_ml_output (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:5270:14)
>   at output_string (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:9322:14)
>   at print_endline (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:9421:7)
>   at caml_call1 (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:247603:28)
>   at Main (/home/mukund/src/tezos-contract/tmp-smartpy-cli/smartpyc.js:248536:9)
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   This happens during the `SmartPy.sh test` step on L107 while trying to construct the contract_*.tar.gz.

Ok, not as familiar but it seems like their installer script might be broken for Nix? Although they do have build instructions from source for NixOS here: https://gitlab.com/SmartPy/smartpy#via-nix 

Check it out and let me know if there are parts of the `env/nix/init` script that we can use?

**Pull request merged**